drivers/dma/dma_iproc_pax_v2.c

```diff
@@ -0,0 +1,940 @@
+	rm_cfg_finish(pd);
+
+#ifndef CONFIG_DMA_IPROC_PAX_POLL_MODE
+	/* Register and enable RM interrupt */
+	IRQ_CONNECT(DT_INST_IRQN(0),
+		    DT_INST_IRQ(0, priority),
+		    rm_isr,
+		    DEVICE_GET(dma_iproc_pax),
+		    0);
+	irq_enable(DT_INST_IRQN(0));
+#else
+	LOG_INF("%s PAX DMA rings in poll mode!\n", DEV_NAME(dev));
+#endif
+	LOG_INF("%s RM setup %d rings\n", DEV_NAME(dev), pd->used_rings);
+
+	return 0;
+}
+
+static int dma_iproc_pax_do_xfer(struct device *dev,
+				 enum ring_idx idx,
+				 struct dma_iproc_pax_payload *pl,
+				 uint32_t pl_len)
+{
+	struct dma_iproc_pax_data *pd = DEV_DATA(dev);
+	struct dma_iproc_pax_cfg *cfg = DEV_CFG(dev);
+	int ret = 0, cnt;
+	struct dma_iproc_pax_rm_ring_data *ring;
+	void *hdr;
+	uint32_t toggle_bit;
+	struct dma_iproc_pax_payload sync_pl;
+	struct dma_iproc_pax_addr64 sync;
+
+	ring = &(pd->ring[idx]);
+	pl = ring->payload;
+
+	/*
+	 * Host sync buffer isn't ready at zephyr/driver init-time
+	 * Read the host address location once at first DMA write
+	 * on that ring.
+	 */
+	if ((ring->sync_pci.addr_lo == 0x0) &&
+	    (ring->sync_pci.addr_hi == 0x0)) {
+		/* populate sync data location */
+		LOG_DBG("sync addr loc 0x%x\n", cfg->scr_addr_loc);
+
+		sync.addr_lo = sys_read32(cfg->scr_addr_loc + 4);
+		sync.addr_hi = sys_read32(cfg->scr_addr_loc);
+		ring->sync_pci.addr_lo = sync.addr_lo + idx * 4;
+		ring->sync_pci.addr_hi = sync.addr_hi;
+		LOG_DBG("ring:%d,sync addr:0x%x.0x%x\n", idx,
+			    ring->sync_pci.addr_hi,
+			    ring->sync_pci.addr_lo);
+	}
+
+	/* account extra sync packet */
+	ring->curr.sync_data.opaque = ring->curr.opq;
+	ring->curr.sync_data.total_pkts = pl_len;
+	memcpy((void *)&ring->sync_loc,
+	       (void *)&(ring->curr.sync_data), 4);
+	sync_pl.pci.addr_lo = ring->sync_pci.addr_lo;
+	sync_pl.pci.addr_hi = ring->sync_pci.addr_hi;
+	sync_pl.axi.addr_lo = (uint32_t)&ring->sync_loc;
+	sync_pl.axi.addr_hi = 0x0; /* always in low mem */
+
+	sync_pl.xfer_sz = 4; /* 4-bytes */
+	sync_pl.direction = CARD_TO_HOST;
+
+	/* Get descriptor write pointer for first header */
+	hdr = (void *)ring->curr.write_ptr;
+	/* current toggle bit */
+	toggle_bit = ring->curr.toggle;
+	/* current opq value for cmpl check */
+	ring->curr.opq = curr_pkt_id(ring);
+	/* Form descriptors for total sg payload */
+	for (cnt = 0; cnt < pl_len; cnt++) {
+		RM_HEADER_V2(__CURR_INC(ring), curr_toggle_val(ring),
+			     curr_pkt_id(ring), RM_DESC_BDCOUNT,
+			     pl+cnt);
+		PCIE_DESC(__CURR_INC(ring), pl + cnt,
+			  curr_toggle_val(ring));
+		SRC_DST_DESC(__CURR_INC(ring), pl + cnt,
+			     curr_toggle_val(ring));
+	}
+	/* Append write sync payload descriptors */
+	RM_HEADER_V2(__CURR_INC(ring), curr_toggle_val(ring),
+		     curr_pkt_id(ring), RM_DESC_BDCOUNT,
+		     &sync_pl);
+	PCIE_DESC(__CURR_INC(ring), &sync_pl,
+		  curr_toggle_val(ring));
+	SRC_DST_DESC(__CURR_INC(ring), &sync_pl,
+		     curr_toggle_val(ring));
+	alloc_pkt_id(ring);
+
+#ifndef CONFIG_DMA_IPROC_PAX_POLL_MODE
```

Comment:
This is probably better in a separate function that's always called, but with an implemention that is an empty static inline if the config option is not set. That way the code flow here doesn't need an ifdef.

Reply:
Sure, wil fix it.